// xmlbuilder2/lib/builder/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.fragmentCB = exports.createCB = exports.convert = exports.fragment = exports.create = exports.builder = exports.XMLBuilderCBImpl = exports.XMLBuilderImpl = void 0;
var XMLBuilderImpl_1 = require("./XMLBuilderImpl");
Object.defineProperty(exports, "XMLBuilderImpl", { enumerable: true, get: function () { return XMLBuilderImpl_1.XMLBuilderImpl; } });
var XMLBuilderCBImpl_1 = require("./XMLBuilderCBImpl");
Object.defineProperty(exports, "XMLBuilderCBImpl", { enumerable: true, get: function () { return XMLBuilderCBImpl_1.XMLBuilderCBImpl; } });
var BuilderFunctions_1 = require("./BuilderFunctions");
Object.defineProperty(exports, "builder", { enumerable: true, get: function () { return BuilderFunctions_1.builder; } });
Object.defineProperty(exports, "create", { enumerable: true, get: function () { return BuilderFunctions_1.create; } });
Object.defineProperty(exports, "fragment", { enumerable: true, get: function () { return BuilderFunctions_1.fragment; } });
Object.defineProperty(exports, "convert", { enumerable: true, get: function () { return BuilderFunctions_1.convert; } });
var BuilderFunctionsCB_1 = require("./BuilderFunctionsCB");
Object.defineProperty(exports, "createCB", { enumerable: true, get: function () { return BuilderFunctionsCB_1.createCB; } });
Object.defineProperty(exports, "fragmentCB", { enumerable: true, get: function () { return BuilderFunctionsCB_1.fragmentCB; } });
