// @oozcitak/infra/lib/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.string = exports.stack = exports.set = exports.queue = exports.namespace = exports.map = exports.list = exports.json = exports.codePoint = exports.byteSequence = exports.byte = exports.base64 = void 0;
const base64 = __importStar(require("./Base64"));
exports.base64 = base64;
const byte = __importStar(require("./Byte"));
exports.byte = byte;
const byteSequence = __importStar(require("./ByteSequence"));
exports.byteSequence = byteSequence;
const codePoint = __importStar(require("./CodePoints"));
exports.codePoint = codePoint;
const json = __importStar(require("./JSON"));
exports.json = json;
const list = __importStar(require("./List"));
exports.list = list;
const map = __importStar(require("./Map"));
exports.map = map;
const namespace = __importStar(require("./Namespace"));
exports.namespace = namespace;
const queue = __importStar(require("./Queue"));
exports.queue = queue;
const set = __importStar(require("./Set"));
exports.set = set;
const stack = __importStar(require("./Stack"));
exports.stack = stack;
const string = __importStar(require("./String"));
exports.string = string;
